### DISCUSSION TITLE
Use of per-asset config in IO managers (new Pythonic types)

### DISCUSSION THREAD

**Discussion by jmccartin** [dagster-io/dagster#19948] [Q&A] [answered]
In our codebase, we have many assets that use a single custom IO manager. When these assets are updated on disk, the IO manager requires knowledge of a given primary key to perform a merge, which can be unique to the asset. This historically has been handled via use of the `config_schema` parameter to each asset, for example (this is greatly simplified):

```python
@asset(
    config_schema=[{"primary_key": Field("foo")}],
)
def generate_asset(upstream: DataFrame) -> DataFrame:
    ...

# IO manager implementation

class MergeIOManager(IOManager):

    @staticmethod
    def _merge_with_key(obj: DataFrame, key: str):
        ...

    def handle_output(self, context: OutputContext, obj: DataFrame) -> None:
        config = context.step_context.op_config
        primary_key = config.get("primary_key")
        self._merge_with_key(obj, primary_key)
```

This means we can instantiate the IO manager once across all assets, and have configuration passed down when the asset is run.

With the new Pythonic resource configuration system however, this does not appear to be possible. From all of the documentation I have read, one must explicitly put all config parameters as class vars to the IO manager (or a subset of them). Something like:

```python
class MergeIOManager(IOManager):

    primary_key: str = pydantic.Field(...)

    @staticmethod
    def _merge_with_key(obj: DataFrame, key: str):
        ...

    def handle_output(self, context: OutputContext, obj: DataFrame) -> None:
        self._merge_with_key(obj, self.primary_key)

# instantiation
merge_io_manager = MergeIOManager(primary_key="foo")
```
As you can see, this means the primary key config var must be declared when the io manager is instantiated. This means that if each asset was to specify this key differently, you would need a separate instantiation for each asset, as there would be no way to read in the config from the example asset below:
```python

class PrimaryKeyConf(Config):
    primary_key: str = pydantic.Field(...)

@asset
def generate_asset(upstream: DataFrame, config: PrimaryKeyConf) -> DataFrame:
    ...
```
Is this assumption correct? If so, this has large refactoring consequences on our existing codebase, so I want to explore other options first. I do not want to remain with the old system if it will be deprecated in the near future.

Finally, how could you pass down run configuration with the new Pythonic type system to an IO manager or resource, if you truly had something that was only known at runtime? Thanks in advance 🙂

**Comment by jamiedemaria** [comment, marked as answer]
hey @jmccartin  you general assessment of this issue is correct. One option could be to set the per-asset information as `metadata` instead of via the `config_schema`. One potential issue with that is that the metadata can't be modified at run time. I'm not sure if you are changing the `primary_key` each time an asset is materialized, or if there is a deterministic mapping between primary key and asset 

```python
@asset(
    metadata={"primary_key": "foo"},
)
def generate_asset(upstream: DataFrame) -> DataFrame:
    ...

class MergeIOManager(IOManager):

    @staticmethod
    def _merge_with_key(obj: DataFrame, key: str):
        ...

    def handle_output(self, context: OutputContext, obj: DataFrame) -> None:
        metadata = context.metadata
        primary_key = metadata.get("primary_key")
        self._merge_with_key(obj, primary_key)
```

**Reply by jmccartin** [reply]
Hi, thanks for the quick reply on this.

I think we looked at using metadata to perform this operation in the past, but it wasn't clear if future Dagster development would have limited values to only certain types. Right now the code suggestion you get works, but mypy complains about `context.metadata` always being `None` (so no values can be received). This can be waived, but is messy. Right now we are looking to extent this passing down of configuration to support a DataFrame schema, and so it's a judgement call whether or not to move to the new system, and the large amount of refactoring that would entail.

Quick question, is it likely that support for the old `config_schema` method will be deprecated in the near future? This helps us inform our decision somewhat. Thanks again.

**Reply by jamiedemaria** [reply]
cc @benpankow  for the future of `config_schema` 

> mypy complains about context.metadata always being None

yeah if there is no metadata attached to the asset, then `context.metadata` would be None, so mypy is correct in flagging this. But it shouldn't *always* be None. If there is metadata on the asset then `context.metadata` will contain the metadata mapping

>  passing down of configuration to support a DataFrame schema

This seems reasonable. In the future we may limit metadata so that you can't pass entire Dataframes, but things like schemas are likely to remain possible
